src/backend/opencl/convolve.cpp

```diff
@@ -85,5 +94,197 @@ INSTANTIATE(ushort ,   float)
 INSTANTIATE(short  ,   float)
 INSTANTIATE(uintl  ,   float)
 INSTANTIATE(intl   ,   float)
+#undef INSTANTIATE
+
+template<typename T, typename accT>
+Array<T> convolve2_unwrap(const Array<T>& signal,
+                          const Array<accT>& filter,
+                          const dim4 stride,
+                          const dim4 padding,
+                          const dim4 dilation)
+{
+    dim4 sDims = signal.dims();
+    dim4 fDims = filter.dims();
+
+    dim_t outputWidth  = 1 + (sDims[0] + 2 * padding[0] - (((fDims[0] - 1) * dilation[0]) + 1)) / stride[0];
+    dim_t outputHeight = 1 + (sDims[1] + 2 * padding[1] - (((fDims[1] - 1) * dilation[1]) + 1)) / stride[1];
+    Array<T> out = createValueArray<T>(dim4(outputWidth, outputHeight, fDims[3], sDims[3]), scalar<T>(0));
+
+    const bool retCols = false;
+    Array<accT> unwrapped = unwrap_dilated(cast<accT>(signal), fDims[0], fDims[1],
+                                                             stride[0], stride[1],
+                                                           padding[0], padding[1],
+                                                         dilation[0], dilation[1], retCols);
+
+    unwrapped = reorder(unwrapped, dim4(1, 2, 0, 3));
+    dim4 uDims = unwrapped.dims();
+    unwrapped.modDims(dim4(uDims[0] * uDims[1], uDims[2] * uDims[3]));
+
+    Array<accT> collapsedFilter = filter;
+
+    vector<af_seq> flip_index(4);
+    af_seq s = {(double)(fDims[0] - 1), 0, -1};
+    flip_index[0] = s;
+    s = {(double)(fDims[1] - 1), 0, -1};
+    flip_index[1] = s;
+    flip_index[2] = af_span;
+    flip_index[3] = af_span;
+
+    collapsedFilter = createSubArray(collapsedFilter, flip_index);
+    collapsedFilter.modDims(dim4(fDims[0] * fDims[1] * fDims[2], fDims[3]));
+
+    Array<accT> res = matmul(collapsedFilter, unwrapped, AF_MAT_TRANS, AF_MAT_NONE);
+    res.modDims(dim4(collapsedFilter.dims()[1], out.dims()[0], out.dims()[1], signal.dims()[3]));
+    out = cast<T>(reorder(res, dim4(1, 2, 0, 3)));
+
+    return out;
+}
+
+
+template<typename T, typename accT>
+Array<T> convolve2(Array<T> const& signal, Array<accT> const& filter,
+                   const dim4 stride, const dim4 padding, const dim4 dilation) {
+    signal.eval();
+    filter.eval();
+
+    Array<T> out  = createEmptyArray<T>(dim4());
+    out = convolve2_unwrap<T, accT>(signal, filter, stride, padding, dilation);
+
```

Comment:
A separate function `convolve2_unwrap` is not necessary if it is only used here. That way you can even avoid the extra call to `createEmptyArray`